### PR TITLE
[WIP] GP functionality

### PR DESCRIPTION
**Pull request by bwengals**
- adds the [FITC approximation](https://papers.nips.cc/paper/3351-the-generalized-fitc-approximation.pdf)
- I used the GPML and GPflow implementations as reference
- new class is called GPfitc at the moment, has the same AP as GP, but with the additional required arg `inducing_points`
- no tests yet since the GP module is WIP
- see [this gist](https://gist.github.com/bwengals/a213680c6ba3b82edf489e58d278e21b) for some usage examples for the time being
- no automatic optimization routine of inducing point locations, needs to be specified by user
- need to make covariance function methods that return just the diagonal, see note on lines 156, 171 of `gp.py` 
- probably will need to change the structure of all this as more GP stuff is added

### CODE REVIEW

**Comment by fonnesbeck** [comment]
Great work.

Instead of having a separate `GPfitc` interface, what about unifying under `GP` and instantiating the approximation if inducing points are passed, and a regular GP otherwise?

File: pymc3/gp/gp.py [hunk 1]
```diff
@@ -3,80 +3,205 @@
...
+    X : array
+        Grid of points to evaluate Gaussian process over.
     mean_func : Mean
         Mean function of Gaussian process
     cov_func : Covariance
         Covariance function of Gaussian process
-    X : array
-        Grid of points to evaluate Gaussian process over. Only required if the 
-        GP is not an observed variable.
     sigma : scalar or array
```
Comment (fonnesbeck) [inline pymc3/gp/gp.py]: Docstring does not match argument name

**Comment by bwengals** [comment]
This might be a good time to see if people are interested in looking this over, regarding the API, functionality, stuff like that.  Not ready for a detailed code review.  Here is a [gist](https://gist.github.com/bwengals/6ca7ee397b78fc23ccfbd94dfc818adf) demoing the broader functionality.  There is still more to be done though!  Here's my necessary/maybe list of things to add at this point, still looking into the maybes, and wouldn't mind input or suggestions of course.

Necessary:
- tests
- docs
- examples

Maybe?:
- Inducing point location optimization
- multiple dependent vars, Grid, Kronecker methods
- Cov functions with compact support / sparse Cholesky
- An inducing point method for the non-conjugate case
- a state space, time series model

File: pymc3/gp/gp.py [hunk 2]
```diff
@@ -120,32 +389,6 @@ def sample_gp(trace, gp, X_values, samples=None, obs_noise=True, model=None, ran
...
-    if chol_const:
-        n = S_post.shape[0]
-        correction = 1e-6 * tt.nlinalg.trace(S_post) * tt.eye(n)
-
-    gp_post = MvNormal.dist(m_post, S_post + correction, shape=Z.shape[0])
-    samples = [gp_post.random(point=trace[idx]) for idx in indices]
-    return np.array(samples)
+    y = [v for v in model.observed_RVs if v.name==gp.name][0]
+    samples = [gp.distribution.random(point=trace[idx], X_values=X_values, y=y, obs_noise=obs_noise) for idx in indices]
+    return np.array(samples)
```
Comment (fonnesbeck) [inline pymc3/gp/gp.py]: Add new line at end of file

File: pymc3/gp/gp.py [hunk 3]
```diff
@@ -68,17 +125,231 @@ def _random(mean, cov, size=None):
...
+        L_B = cholesky(tt.eye(self.nu) + tt.dot(A_l, tt.transpose(A)))
+        r = y - self.m(self.X)
+        r_l = r / Lamd
+        c = solve_lower(L_B, tt.dot(A, r_l))
+        Kus = self.K(self.Xu, Xs)
+        As = solve_lower(Luu, Kus)
+        mean = tt.dot(tt.transpose(As), solve_upper(tt.transpose(L_B), c))
+        C = solve_lower(L_B, As)
+        if obs_noise:
+            cov = self.K(Xs, Xs) - tt.dot(tt.transpose(As), As)\
```
Comment (fonnesbeck) [inline pymc3/gp/gp.py:357]: Avoid continuation character throughout. Enclose selection in parentheses instead.

**Comment by junpenglao** [comment]
This looks really nice but I am not sure I understand what are the approx do inside of the GP class. Are they operated on the logp to turn it into something easier to approximate/sample from? It is not an estimator right (i.e., I cannot use it to get estimation of parameters)

File: pymc3/gp/gp.py [hunk 4]
```diff
@@ -68,17 +125,231 @@ def _random(mean, cov, size=None):
...
-              chol_const=True):
+    Parameters
+    ----------
+    X : array
+        Array of predictor variables.
+    mean_func : Mean
+        Mean function of Gaussian process
+    cov_func : Covariance
+        Covariance function of Gaussian process
+    sigma2 : scalar or array
```
Comment (fonnesbeck) [inline pymc3/gp/gp.py]: Was wondering about the decision to use sigma2 versus sigma. Would we not be putting a prior on the latter?

File: pymc3/gp/gp.py [hunk 5]
```diff
@@ -3,60 +3,117 @@
...
-    """Gausian process
+def stabilize(K):
+    if CHOL_CONST:
+        n = K.shape[0]
+        return K + 1e-6 * (tt.nlinalg.trace(K)/n) * tt.eye(n)
+    else:
+        return K
 
+
+def GP(name, X, mean_func=None, cov_func=None,
```
Comment (fonnesbeck) [inline pymc3/gp/gp.py]: I like the refactoring

**Comment by fonnesbeck** [comment]
@junpenglao its the sparse approximation method used when inducing points are passed.

File: pymc3/gp/gp.py [hunk 6]
```diff
@@ -3,60 +3,117 @@
...
 
-        if X is not None:
-            self.X = X
-            self.mean = self.mode = self.M(X)
-            kwargs.setdefault("shape", X.squeeze().shape)
+    if inducing_points is None and n_inducing is None:
+        raise ValueError("Must specify one of 'inducing_points' or 'n_inducing'")
 
-        super(GP, self).__init__(*args, **kwargs)
+    if inducing_points is None:
```
Comment (fonnesbeck) [inline pymc3/gp/gp.py]: Can you just nest this condition in the one below

File: pymc3/gp/gp.py [hunk 7]
```diff
@@ -3,60 +3,117 @@
...
+        if sigma2 is None and cov_func_noise is None:
+            raise ValueError('Provide a value or a prior for the noise variance')
+        if sigma2 is not None and cov_func_noise is None:
+            cov_func_noise = lambda X: sigma2 * tt.eye(X.shape[0])
+        return GPFullConjugate(name, X, mean_func, cov_func, cov_func_noise, observed=observed)
+    else:
+        approx = approx.upper()
+
+    ##### CONJUGATE, APPROXIMATION
+    if approx not in ["VFE", "FITC"]:
```
Comment (fonnesbeck) [inline pymc3/gp/gp.py]: What happens if inducing points are passed but approx is None? Should you perhaps just set one as the default?

File: pymc3/gp/gp.py (hunk 4 above)
Comment (bwengals) [reply pymc3/gp/gp.py]: i hadnt really thought about it.  thinking about it more, yeah I think your right it makes more sense to parameterize in terms of sigma, not sigma2

File: pymc3/gp/gp.py (hunk 7 above)
Comment (bwengals) [reply pymc3/gp/gp.py]: good catch, definitely

File: pymc3/gp/gp.py (hunk 3 above)
Comment (bwengals) [reply pymc3/gp/gp.py:357]: will do.  curious though, is it better for readability or?

File: pymc3/gp/gp.py [hunk 8]
```diff
@@ -2,60 +2,124 @@
...
+        The number of inducing points to use
+    inducing_points : array
+        The inducing points to use
+    observed : None or array
+        The observed 'y' values, use if likelihood is Gaussian
+    chol_const: boolean
+        Whether or not to stabilize Cholesky decompositions
     """
-    def __init__(self, mean_func=None, cov_func=None, X=None, sigma=0, *args, **kwargs):
+    global CHOL_CONST
```
Comment (AustinRochford) [inline pymc3/gp/gp.py]: `global` is usually a code smell to me.  Can this be passed as a keyword argument to functions that need it instead?

File: pymc3/gp/gp.py [hunk 9]
```diff
@@ -2,60 +2,124 @@
...
+    else:
+        if not isinstance(mean_func, Mean):
+            raise ValueError('mean_func must be a subclass of Mean')
+    if cov_func is None:
+        raise ValueError('A covariance function must be specified for GPP')
+    if not isinstance(cov_func, Covariance):
+        raise ValueError('cov_func must be a subclass of Covariance')
+
+    # NONCONJUGATE
+    if observed is None:
```
Comment (AustinRochford) [inline pymc3/gp/gp.py]: Unobserved isn' necessarily non-conjugate, or am I missing something?

File: pymc3/gp/gp.py [hunk 10]
```diff
@@ -377,18 +387,30 @@ def sample_gp(trace, gp, X_values, samples=None, obs_noise=True, model=None, ran
...
 
     if random_seed:
         np.random.seed(random_seed)
 
-    indices = np.random.randint(0, len(trace), samples)
+    if samples is None:
+        samples = len(trace)
+
+    indices = np.random.choice(np.arange(samples), len(trace), replace=False)
     if progressbar:
```
Comment (ferrine) [inline pymc3/gp/gp.py]: Tqdm has kwarg `disable=True` for that

Comment (bwengals) [reply pymc3/gp/gp.py]: oh nice, ill use that

File: pymc3/gp/gp.py (hunk 8 above)
Comment (bwengals) [reply pymc3/gp/gp.py]: it could, i thought the global would be cleaner in this case.  only one function `stabilize` uses this keyword, and stabilize is used by most everything.  IMHO `chol_const` doesn't need to be accessed by the user

File: pymc3/gp/gp.py (hunk 9 above)
Comment (bwengals) [reply pymc3/gp/gp.py]: hmm.. if the gp + noise values are observed, and the noise is gaussian, it's best to use conjugacy to integrate out the actual gp so it doesnt need to be sampled.  But thinking about it more, thats only true **if** nothing happens in the model later that depends on the observed gp + noise data.   This would be kind of weird... but possible I guess.

File: pymc3/gp/gp.py (hunk 8 above)
Comment (ferrine) [reply pymc3/gp/gp.py]: There is no need in global as CHOL_CONST is in outer scope or I miss something

Comment (ferrine) [reply pymc3/gp/gp.py]: I've really missed some lines after, sorry. But why do you change global state? Is it so essential? I feel worried too

Comment (bwengals) [reply pymc3/gp/gp.py]: yeah fair enough, it is a bit smelly.  I'll take this out

File: pymc3/gp/gp.py (hunk 9 above)
Comment (AustinRochford) [reply pymc3/gp/gp.py]: Good point. I don't think it's necessary to marginalize here given the rarity of the use case, but perhaps alter the naming?

File: pymc3/gp/gp.py [hunk 11]
```diff
@@ -63,56 +60,64 @@ def GP(name, X, mean_func=None, cov_func=None,
...
+                                   observed=observed)
+        else:
+            # use an approximation
+            if approx is None:
+                pm._log.info("Using VFE approximation")
+                approx = "VFE"
+
+            approx = approx.upper()
+            if approx not in ["VFE", "FITC"]:
+                raise ValueError(("'FITC' or 'VFE' are the implemented "
```
Comment (ferrine) [inline pymc3/gp/gp.py]: better message should say that `'FITC' or 'VFE'` are the only approximations that are implemented and show what has user provided. e.g. `"'FITC' or 'VFE' are the only supported approximations for GP, got %s"`

File: pymc3/gp/gp.py [hunk 12]
```diff
@@ -63,56 +60,64 @@ def GP(name, X, mean_func=None, cov_func=None,
...
+
+            if inducing_points is None and n_inducing is None:
+                raise ValueError(("Must specify one of 'inducing_points' "
+                                  "or 'n_inducing'"))
+
+            if inducing_points is None and n_inducing is not None:
+                # initialize inducing points with K-means
+                from scipy.cluster.vq import kmeans
+                # first whiten X
+                if not isinstance(X, np.ndarray):
```
Comment (ferrine) [inline pymc3/gp/gp.py]: I would rather use `np.asarray(X)` here instead of `X.value`

Comment (ferrine) [reply pymc3/gp/gp.py]: Additionally you can check if X is not a tensor, that can be not a rare case.

File: pymc3/gp/gp.py (hunk 3 above)
Comment (fonnesbeck) [reply pymc3/gp/gp.py:357]: Partly readability, partly convention.

File: pymc3/gp/gp.py (hunk 11 above)
Comment (bwengals) [reply pymc3/gp/gp.py]: good call, changed

File: pymc3/gp/gp.py (hunk 12 above)
Comment (bwengals) [reply pymc3/gp/gp.py]: thanks, the check should be a bit more robust now

**Comment by bwengals** [comment]
@fonnesbeck @AustinRochford @otherinterestedfolks I think this PR is in a state where it's ready for review!  I'll start on the docs next.  I'll submit doc changes to this PR since they will probably help with the review.

File: pymc3/gp/cov.py [hunk 13]
```diff
@@ -35,26 +37,29 @@ def __init__(self, input_dim, active_dims=None):
             if len(active_dims) != input_dim:
                 raise ValueError("Length of active_dims must match input_dim")
 
-    def __call__(self, X, Z=None, diag=False):
+    def __call__(self, X, Xs=None, diag=False):
```
Comment (fonnesbeck) [inline pymc3/gp/cov.py]: Z seemed simpler and more distinct. Why the change?

File: pymc3/gp/gp.py [hunk 14]
```diff
@@ -2,60 +2,140 @@
...
+    mean_func : Mean
+        Mean function of Gaussian process
+    cov_func_noise : Covariance
+        Covariance function of noise process (ignored for approximations)
     sigma : scalar or array
-        Observation standard deviation (defaults to zero)
+        Noise standard deviation
+    approx : None or string,
+        Allowed values are 'FITC' and 'VFE'
+    n_inducing : integer
```
Comment (fonnesbeck) [inline pymc3/gp/gp.py]: Can’t `n_inducing` be deduced from `inducing_points`?

File: pymc3/gp/cov.py (hunk 13 above)
Comment (bwengals) [reply pymc3/gp/cov.py]: Z seemed most commonly used to represent inducing point locations in the lit.  I changed it to Xs to look more like $X_*$

File: pymc3/gp/cov.py [hunk 15]
```diff
@@ -104,17 +105,34 @@ def __init__(self, factor_list):
...
-                if diag:
-                    factor_list.append(tt.diag(factor))
-                else:
-                    factor_list.append(factor)
-            else:
-                factor_list.append(factor)
+                continue
+
+            # if factor is a numpy array
+            if isinstance(factor, np.ndarray):
```
Comment (ferrine) [inline pymc3/gp/cov.py]: I think `elif` is more suitable here

File: pymc3/gp/cov.py [hunk 16]
```diff
@@ -207,6 +205,21 @@ def full(self, X, Xs=None):
         raise NotImplementedError
 
 
+class Periodic(Stationary):
```
Comment (fonnesbeck) [inline pymc3/gp/cov.py:208]: Does this obviate the need for `Cosine`?

File: pymc3/gp/gp.py [hunk 17]
```diff
@@ -411,10 +410,6 @@ def sample_gp(trace, gp, X_new, n_samples=None, obs_noise=True,
     obs_noise : bool
         Flag for including observation noise in sample.  Does not apply to GPs
         used with non-conjugate likelihoods.  Defaults to False.
-    cov_func : Covariance
```
Comment (fonnesbeck) [inline pymc3/gp/gp.py]: Why are the docstrings removed when the arguments are still available?

File: pymc3/gp/cov.py (hunk 16 above)
Comment (bwengals) [reply pymc3/gp/cov.py:208]: I think they can both be here.  Functions drawn from a gp w/ cosine prior are exactly sinusoidal functions.  Periodic, with its lengthscale parameter, gives functions that are periodic but not perfectly sinusoidal.

File: pymc3/gp/gp.py (hunk 17 above)
Comment (bwengals) [reply pymc3/gp/gp.py]: Ah, the arguments should also be gone.  This is where I ran into problems trying to support additive gps.  I'm submitting a new pr with pretty big refactor soon...
